[cmake] use -Werror for double-promotion, missing-field-initializers, and sign-compare

lrusak opened the pull request:
We seem to continue adding warnings and continue having to fix them. Let's make these `-Werror` now to avoid adding new warnings.

I'm not sure if there is any negative repercussions at this time.

I had to disable the behaviour for a couple of the external libraries. This seems like an ok solution for now.

Note that this is only for posix platforms. I'm not sure what the windows changes would be.

lrusak committed:
CAEBitstreamPacker: fix sign-compare warning

xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp:205:24: warning: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘const int’ [-Wsign-compare]
  205 |     if (m_trueHDframes == nFrames)
      |         ~~~~~~~~~~~~~~~^~~~~~~~~~

lrusak committed:
CMediaCodecDecoderFilterManager: fix sign-compare

xbmc/platform/android/media/decoderfilter/MediaCodecDecoderFilterManager.cpp:43:21: error: comparison of integers of different signs: 'int' and 'unsigned int' [-Werror,-Wsign-compare]
  for (int i = 0; i < num_codecs; i++)
                  ~ ^ ~~~~~~~~~~

AlwinEsch commented:
I think it is good to use more warnings and it makes the code more secure, especially since gcc 11 itself already has a higher definition and is then noticed more quickly on older ones.

I'm just a little torn because of the depends, I think it's great to see warnings there because you may then recognize background problems. Stupid only that with some depends a "no-error" is necessary, hope is not lost with updates to check.

I wouldn't worry that much about Windows, because the warning declaration is very high there anyway.

lrusak committed:
[cmake] allow specifying compile options for core units only

lrusak committed:
[cmake] clean up compile options

lrusak committed:
[cmake] add -Wsign-compare warning

lrusak commented:
I went a bit of a different direction with this. We can't really use the cmake `add_options` like we do for warnings as they apply to all code built by cmake (including cmake external projects, some dependencies, and the lib/ folder).

I've added a couple comments about where options apply so when adding new ones.

The flags now look like this for me. Note that later options override earlier ones.
```
CXX_FLAGS = -Wall -Wdouble-promotion -Wmissing-field-initializers -Wsign-compare -g -D_DEBUG <snip> -Werror=double-promotion -Werror=missing-field-initializers -Werror=sign-compare -std=c++14
```

lrusak committed:
CDVDAudioCodecAndroidMediaCodec: fix sign-compare warning

xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecAndroidMediaCodec.cpp:205:23: error: comparison of integers of different signs: 'int' and 'unsigned int' [-Werror,-Wsign-compare]
   for (int i = 0; i < num_codecs; i++)
                   ~ ^ ~~~~~~~~~~

lrusak committed:
CWinSystemOSX: fix sign-compare warning

xbmc/windowing/osx/SDL/WinSystemOSXSDL.mm:210:41: error: comparison of integers of different signs: 'int' and 'CGDisplayCount' (aka 'unsigned int') [-Werror,-Wsign-compare]
  if (screen_index >= 0 && screen_index < numDisplays)
                           ~~~~~~~~~~~~ ^ ~~~~~~~~~~~

lrusak committed:
CDVDVideoCodecDRMPRIME: fix double-promotion warning

xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp:478:51: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
        static_cast<float>(pVideoPicture->iWidth) / static_cast<float>(pVideoPicture->iHeight);
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lrusak commented:
jenkins build this please

wsnipex commented:
looks sane to me, only question is if we should make CORE_COMPILE_OPTIONS an actual option or user overridable.
Did you test, if setting CXXFLAGS or CMAKE_CXX_FLAGS allows turning those off?

lrusak committed:
CColorManager: fix double-promotion warning

xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:492:44: error: implicit conversion from ‘float’ to ‘double’ to match other operand of binary expression [-Werror=double-promotion]
  492 |       double bkipow = pow(blackPoint.Y, 1.0/gammaValue);
      |                                         ~~~^~~~~~~~~~~

lrusak commented:
> looks sane to me, only question is if we should make CORE_COMPILE_OPTIONS an actual option or user overridable. Did you test, if setting CXXFLAGS or CMAKE_CXX_FLAGS allows turning those off?

That's a good point. I'm not sure what the best way to handle that is. It seems that using `CXXFLAGS` env variable or `-DCMAKE_CXX_FLAGS` configure variable both prepend the flags. That won't work as latter options override prior options.

So that maybe makes it an options that can be overridden (although I don't like adding new options). Do we just leave it as is for now and see if it becomes problematic?

wsnipex commented on cmake/scripts/common/ArchSetup.cmake:
```diff
@@ -148,8 +148,24 @@ if(NOT DEFINED NEON OR NEON)
+  )
+
+  # these options affect only core code
+  set(CORE_COMPILE_OPTIONS
```
let's allow the user to override:
```suggestion
  if(NOT CORE_COMPILE_OPTIONS)
    set(CORE_COMPILE_OPTIONS
```

should be sufficient, without adding an explicit option

lrusak replied:
I'm fine with that 👍

lrusak committed:
CCPUInfoFreebsd: fix sign-compare warning

xbmc/platform/freebsd/CPUInfoFreebsd.cpp:63:24: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  for (size_t i = 0; i < m_cpuCount; i++)
                     ~ ^ ~~~~~~~~~~

xbmc/platform/freebsd/CPUInfoFreebsd.cpp:192:24: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  for (size_t i = 0; i < m_cpuCount; i++)
                     ~ ^ ~~~~~~~~~~

lrusak committed:
CCPUInfoFreebsd: fix double-promotion warning

xbmc/platform/freebsd/CPUInfoFreebsd.cpp:225:56: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
    m_cores[core].m_usagePercent = activeTime * 100.0f / totalTime;
                                 ~ ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~

lrusak commented:
squashed relevant fixups are reorder the commits to have all the fix commits before enabling `-Werror`

AlwinEsch commented:
jenkins build this with addons please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

AlwinEsch commented:
Have you somewhere the link about Jenkins and binary addons, want to make sure that it not affects them?

mkreisl commented:
```
grep warning.*implicit build.log | wc -l
66
```
(32bit Debian Bullseye build for Raspberry Pi4)
This commit is once again a good example of a senseless waste of time and resources here.
 I'm sorry, but I had to say that.

thexai commented:
> We seem to continue adding warnings and continue having to fix them.

Yeah!
https://github.com/xbmc/xbmc/commit/76105534ad94d9b25832520231f615cc4a6109fb has caused 4 **new** Visual Studio warnings:

https://jenkins.kodi.tv/view/Windows/job/WIN-64/19260/msbuild/folder.917671344/

https://jenkins.kodi.tv/view/Windows/job/WIN-64/19260/msbuild/folder.917671344/source.9082885b-d4cb-4305-a96a-91e286d7455b/#57

lrusak commented:
> Have you somewhere the link about Jenkins and binary addons, want to make sure that it not affects them?

@AlwinEsch 
https://jenkins.kodi.tv/view/Automation/job/BuildMultiWithAddons-PR/lastFailedBuild/

lrusak commented:
> ```
> 
> grep warning.*implicit build.log | wc -l
> 
> 66
> 
> ```
> 
> (32bit Debian Bullseye build for Raspberry Pi4)
> 
> This commit is once again a good example of a senseless waste of time and resources here.
> 
>  I'm sorry, but I had to say that.

I'm a little lost. What is the point of this comment?

lrusak commented:
jenkins build this please

lrusak commented:
Let's see what happens here.... 🙃

howie-f commented:
_dang_
```
GUIFontTTF.cpp: In member function 'std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText&)':
GUIFontTTF.cpp:666:49: error: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Werror=sign-compare]
  666 |       for (unsigned int j = lastSetIndex + 1; j < i; j++)
      |                                               ~~^~~
cc1plus: some warnings being treated as errors
```

Portisch commented:
```
/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1173:44: error: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Werror=sign-compare]
 1173 |     (stream->disabled && stream->ExtraSize != m_pFormatContext->streams[pPacket->iStreamId]->codecpar->extradata_size))
      |                          ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
